src/hotspot/cpu/riscv/vm_version_riscv.hpp

```diff
@@ -30,55 +31,174 @@
 #include "runtime/abstract_vm_version.hpp"
 #include "runtime/arguments.hpp"
 #include "runtime/globals_extension.hpp"
+#include "utilities/growableArray.hpp"
 #include "utilities/sizes.hpp"
 
+class RiscvHwprobe;
+
 class VM_Version : public Abstract_VM_Version {
-#ifdef COMPILER2
-private:
-  static void c2_initialize();
-#endif // COMPILER2
+  friend RiscvHwprobe;
+ private:
+  class RVFeatureValue {
+    const char* _pretty;
+    bool        _feature_string;
+    uint64_t    _feature_bit;
+    bool        _enabled;
+    int64_t     _value;
+   public:
+    RVFeatureValue(const char* pretty, uint64_t bit, bool fstring) :
+      _pretty(pretty), _feature_string(fstring), _feature_bit(0),
+      _enabled(false), _value(-1) {
+        _feature_bit = (1ULL << bit);
+    }
+    void enable_feature(int64_t value = 0) {
+      _enabled = true;
+      _value = value;
+    }
+    const char* pretty()   { return _pretty; }
+    uint64_t feature_bit() { return _feature_bit; }
+    bool feature_string()  { return _feature_string; }
+    bool enabled()         { return _enabled; }
+    int64_t value()        { return _value; }
+    virtual void update_flag() = 0;
+  };
 
-// VM modes (satp.mode) privileged ISA 1.10
-enum VM_MODE {
-  VM_MBARE = 0,
-  VM_SV39  = 8,
-  VM_SV48  = 9,
-  VM_SV57  = 10,
-  VM_SV64  = 11
-};
+  #define UPDATE_DEFAULT(flag)        \
+  void update_flag() {                \
+      assert(enabled(), "Must be.");  \
+      if (FLAG_IS_DEFAULT(flag)) {    \
+        FLAG_SET_DEFAULT(flag, true); \
+      }                               \
+  }                                   \
+
+  #define NO_UPDATE_DEFAULT           \
+  void update_flag() {}               \
+
+  // Frozen standard extensions
+  // I RV64I
+  // M Integer Multiplication and Division
+  // A Atomic Instructions
+  // F Single-Precision Floating-Point
+  // D Single-Precision Floating-Point
+  // (G = M + A + F + D)
+  // Q Quad-Precision Floating-Point
+  // C Compressed Instructions
+  // H Hypervisor
+  //
+  // Others, open and non-standard
+  // V Vector
+  //
+  // Cache Management Operations
+  // Zicbom Cache Block Management Operations
+  // Zicboz Cache Block Zero Operations
+  // Zicbop Cache Block Prefetch Operations
+  //
+  // Bit-manipulation
+  // Zba Address generation instructions
+  // Zbb Basic bit-manipulation
+  // Zbc Carry-less multiplication
+  // Zbs Single-bit instructions
+  //
+  // Zicsr Control and Status Register (CSR) Instructions
+  // Zifencei Instruction-Fetch Fence
+  // Zic64b Cache blocks must be 64 bytes in size, naturally aligned in the address space.
+  // Zihintpause Pause instruction HINT
+  //
+  // Other features and settings
+  // mvendorid Manufactory JEDEC
+  // marchid   Manufactory prop.
+  // mimpid    Manufactory prop.
```

Comment:
I think the code comments for marchid and mimpid could be improved.

Reply:
Ok, I'll fix something!